### PR TITLE
polish code and add multi-card validate and infer in image classification

### PR DESCRIPTION
**Pull request by shippingwang**
- 预测一张图脚本
指定 --image_path="XXX"
如python infer.py --model=ResNet50 --pretrained_model="XXX/" --sample_path="XXX.jpeg" 
image

- 预测结果保存json 
python infer.py --save_json="xxx.json"

- 评估脚本, 结果输出到json文件中           
指定 json文件保存地址 --save_json="XXX"

image

- 模型转换脚本
还是infer.py指定--save_inference来做

- 没有校验集合，只支持训练，训练过程中不评估
指定--validate=False即可

- 模型微调脚本
去除FC层
- 分类: 代码包里写死了acc5的计算，用户设置类别数class_dim小于5的时候会报错

- 支持多卡评估 ，单独评估&边训边测，支持多卡预测

### CODE REVIEW

File: PaddleCV/image_classification/build_model.py [hunk 1]
```diff
@@ -48,7 +48,8 @@ def _basic_model(data, model, args, is_train):
 
     avg_cost = fluid.layers.mean(cost)
     acc_top1 = fluid.layers.accuracy(input=softmax_out, label=label, k=1)
-    acc_top5 = fluid.layers.accuracy(input=softmax_out, label=label, k=5)
+    acc_top5 = fluid.layers.accuracy(
+        input=softmax_out, label=label, k=min(5, args.class_dim))
```
Comment (dyning) [inline PaddleCV/image_classification/build_model.py:52]: 是不是需要假设class_dim>1?

File: PaddleCV/image_classification/build_model.py [hunk 2]
```diff
@@ -73,7 +74,8 @@ def _googlenet_model(data, model, args, is_train):
 
     avg_cost = avg_cost0 + 0.3 * avg_cost1 + 0.3 * avg_cost2
     acc_top1 = fluid.layers.accuracy(input=out0, label=label, k=1)
-    acc_top5 = fluid.layers.accuracy(input=out0, label=label, k=5)
+    acc_top5 = fluid.layers.accuracy(
+        input=out0, label=label, k=min(5, args.class_dim))
```
Comment (dyning) [inline PaddleCV/image_classification/build_model.py:78]: 与上面类似

File: PaddleCV/image_classification/eval.py [hunk 3]
```diff
@@ -48,9 +49,16 @@
 add_arg('interpolation',    int,  None,                 "The interpolation mode")
 add_arg('padding_type',     str,  "SAME",               "Padding type of convolution")
 add_arg('use_se',           bool, True,                 "Whether to use Squeeze-and-Excitation module for EfficientNet.")
+add_arg('json',             str,  None,                 "Whether to save output in json file.")
```
Comment (dyning) [inline PaddleCV/image_classification/eval.py]: 这个字段意义不明，是保存批量预测结果还是保存评测结果？

File: PaddleCV/image_classification/infer.py [hunk 4]
```diff
@@ -46,10 +46,11 @@
 parser.add_argument('--image_std', nargs='+', type=float, default=[0.229, 0.224, 0.225], help="The std of input image data")
 parser.add_argument('--image_shape', nargs='+', type=int, default=[3, 224, 224], help="the shape of image")
 add_arg('topk',             int,  1,                    "topk")
-add_arg('label_path',       str,  "./utils/tools/readable_label.txt", "readable label filepath")
+add_arg('map_path',       str,  "./utils/tools/readable_label.txt", "readable label filepath")
```
Comment (dyning) [inline PaddleCV/image_classification/infer.py]: 名称建议改为cls_map_path

File: PaddleCV/image_classification/infer.py [hunk 5]
```diff
@@ -46,10 +46,11 @@
...
 add_arg('interpolation',    int,  None,                 "The interpolation mode")
 add_arg('padding_type',     str,  "SAME",               "Padding type of convolution")
 add_arg('use_se',           bool, True,                 "Whether to use Squeeze-and-Excitation module for EfficientNet.")
+add_arg('sample_path',      str,  None,                 "single sample image path")
```
Comment (dyning) [inline PaddleCV/image_classification/infer.py]: 建议改为image_path

File: PaddleCV/image_classification/utils/utility.py [hunk 6]
```diff
@@ -147,7 +147,8 @@ def parse_args():
     add_arg('is_profiler',              int,    0,                      "the profiler switch.(used for benchmark)")
     add_arg('profiler_path',            str,    './',                   "the profiler output file path.(used for benchmark)")
     add_arg('max_iter',                 int,    0,                    "the max train batch num.(used for benchmark)")
-    add_arg('validate',                 int,    1,                      "whether validate.(used for benchmark)")
+
+    add_arg('validate',                 int,    1,                      "whether to validate when training.")
```
Comment (dyning) [inline PaddleCV/image_classification/utils/utility.py]: flag_uesed_validate

Comment (shippingwang) [reply PaddleCV/image_classification/utils/utility.py]: 这个是做QA做benchmark的加的，我把这个写完善了，应该叫这个名字统一些。

File: PaddleCV/image_classification/infer.py (hunk 5 above)
Comment (shippingwang) [reply PaddleCV/image_classification/infer.py]: OK

File: PaddleCV/image_classification/infer.py (hunk 4 above)
Comment (shippingwang) [reply PaddleCV/image_classification/infer.py]: ok

File: PaddleCV/image_classification/eval.py (hunk 3 above)
Comment (shippingwang) [reply PaddleCV/image_classification/eval.py]: eval和infer文件是分开的，在eval.py里的是评测结果，改成了save_json_path

File: PaddleCV/image_classification/build_model.py (hunk 2 above)
Comment (shippingwang) [reply PaddleCV/image_classification/build_model.py:78]: OK

File: PaddleCV/image_classification/build_model.py (hunk 1 above)
Comment (shippingwang) [reply PaddleCV/image_classification/build_model.py:52]: OK

File: PaddleCV/image_classification/utils/utility.py [hunk 7]
```diff
@@ -468,7 +493,8 @@ def best_strategy_compiled(args, program, loss, exe):
             exec_strategy.num_threads = 1
 
         compiled_program = fluid.CompiledProgram(program).with_data_parallel(
-            loss_name=loss.name,
+            loss_name=loss.name if mode == "train" else loss,
+            share_vars_from=share_prog is mode == "val",
```
Comment (LielinJiang) [inline PaddleCV/image_classification/utils/utility.py]: confused with this line

Comment (shippingwang) [reply PaddleCV/image_classification/utils/utility.py]: I don't change the parameter list of validate function, 
it passes a loss_name(test_fetch_list[0]) not the loss var(test_fetch_vars[0]).
so add a branch here...

**Review by heavengate — APPROVED** [review]
